asmcomp/selectgen.ml

```diff
@@ -194,18 +195,139 @@ let join_array rs =
 (* Name of function being compiled *)
 let current_function_name = ref ""
 
+(* Environment parameter for the function being compiled, if any. *)
+let current_function_env_param = ref None
```

Comment:
The role of this reference is to detect a common case of reading from an immutable block (the current closure).  This seems a bit hackish.  Couldn't one mark the load instruction instead?  This seems cleaner and might (perhap later) cover more cases.